apps/meteor/app/lib/server/functions/setUsername.ts

```diff
@@ -3,15 +3,70 @@ import { Accounts } from 'meteor/accounts-base';
 import type { IUser } from '@rocket.chat/core-typings';
 import { Invites, Users } from '@rocket.chat/models';
 import { api } from '@rocket.chat/core-services';
+import _ from 'underscore';
 
 import { settings } from '../../../settings/server';
 import { hasPermissionAsync } from '../../../authorization/server/functions/hasPermission';
 import { RateLimiter } from '../lib';
 import { addUserToRoom } from './addUserToRoom';
-import { setUserAvatar } from '.';
+import { saveUserIdentity, setUserAvatar } from '.';
 import { checkUsernameAvailability } from './checkUsernameAvailability';
 import { getAvatarSuggestionForUser } from './getAvatarSuggestionForUser';
 import { SystemLogger } from '../../../../server/lib/logger/system';
+import { callbacks } from '../../../../lib/callbacks';
+import { joinDefaultChannels } from './joinDefaultChannels';
+
+export const setUsernameWithValidation = async (userId: string, username: string, joinDefaultChannelsSilenced?: boolean): Promise<void> => {
+	if (!username) {
+		throw new Meteor.Error('error-invalid-username', 'Invalid username', { method: 'setUsername' });
+	}
+
+	const user = await Users.findOneById(userId);
```

Comment:
Since you're not using the full user, a projection would be good :)

Reply:
My main concern here is the callback.run we have below, it's relying on the whole user 😞